tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+package at.forsyte.apalache.tla.bmcmt.smt
+
+import at.forsyte.apalache.tla.bmcmt.profiler.SmtListener
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell}
+import at.forsyte.apalache.tla.lir.TlaEx
+import com.typesafe.scalalogging.LazyLogging
+
+
+object RecordingZ3SolverContext {
+  def apply(parentLog: Option[SmtLog], config: SolverConfig): RecordingZ3SolverContext = {
+    val context = new RecordingZ3SolverContext(parentLog, config)
+    parentLog match {
+      case Some(log) =>
+        log.replay(context.solver)
+
+      case None => ()
+    }
+    context
+  }
+}
+
+/**
+  * A wrapper around SolverContext that records SMT operations. The log is organized hierarchically,
+  * so a differential diff may be taken.
+  *
+  * @param parentLog the log of the parent context
+  * @param config solver configuration
+  *
+  * @author Igor Konnov
+  */
+@SerialVersionUID(700L)
+class RecordingZ3SolverContext private (parentLog: Option[SmtLog], val config: SolverConfig)
+    extends SolverContext with Serializable with LazyLogging {
+
+  import SmtLog._
+
+  private val solver = new Z3SolverContext(config)
+
+  /**
+    * The sequence of logs (the last added element is in the head),
+    * one per context, except the last log, which is maintained in lastLog.
+    * Every list in logStackRev is stored in the reverse order.
+    */
+  private var logStackRev: List[List[Record]] = List()
+
+  /**
+    * The current log, the last added element is in the head.
+    */
+  private var lastLogRev: List[Record] = List()
```

Comment:
Same nit re: naming here.

Is this `lastLog` meaning the most recent log, that should be the head of `logStack`? If so, why do we keep that in a separate variable? Getting the head of the `logStack` should be trivial, right? If we do indeed need this pointer to the head element for some reason, I think a line or two up here explaining to the reader *why* we need it would be helpful.

Reply:
It is hard to justify why I did it like that. It was months ago. I am going to refactor the code as you are suggesting. Good catch!